Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean

```diff
@@ -0,0 +1,113 @@
+
+/-!
+
+# Residue fields of points
+
+Any point `x` of a locally ringed space `X` comes with a natural residue field, namely the residue
+field of the stalk at `x`. Moreover, for every open subset of `X` containing `x`, we have a
+canonical evaluation map from `Γ(X, U)` to the residue field of `X` at `x`.
+
+## Main definitions
+
+The following are in the `AlgebraicGeometry.LocallyRingedSpace` namespace:
+
+- `residueField`: the residue field of the stalk at `x`.
+- `evaluation`: for open subsets `U` of `X` containing `x`, the evaluation map from sections over
+  `U` to the residue field at `x`.
+- `evaluationMap`: a morphism of locally ringed spaces induces a morphism, i.e. extension, of
+  residue fields.
+
+-/
+
+universe u
+
+open CategoryTheory TopologicalSpace Opposite
+
+noncomputable section
+
+namespace AlgebraicGeometry.LocallyRingedSpace
+
+variable (X : LocallyRingedSpace.{u}) {U : Opens X}
+
+/-- The residue field of `X` at a point `x` is the residue field of the stalk of `X`
+at `x`. -/
+def residueField (x : X) : CommRingCat :=
+  CommRingCat.of <| LocalRing.ResidueField (X.stalk x)
+
+instance (x : X) : Field (X.residueField x) :=
+  inferInstanceAs <| Field (LocalRing.ResidueField (X.stalk x))
+
+/--
+If `U` is an open of `X` containing `x`, we have a canonical ring map from the sections
+over `U` to the residue field of `x`.
+
+If we interpret sections over `U` as functions of `X` defined on `U`, then this ring map
+corresponds to evaluation at `x`.
+-/
+def evaluation (x : U) : X.presheaf.obj (op U) ⟶ X.residueField x :=
+  X.presheaf.germ x ≫ LocalRing.residue _
+
+/-- The global evaluation map from `Γ(X, ⊤)` to the residue field at `x`. -/
+def Γevaluation (x : X) : X.presheaf.obj (op ⊤) ⟶ X.residueField x :=
+  X.evaluation ⟨x, show x ∈ ⊤ from trivial⟩
+
+@[simp]
+lemma evaluation_eq_zero_iff_not_mem_basicOpen (x : U) (f : X.presheaf.obj (op U)) :
+    X.evaluation x f = 0 ↔ x.val ∉ X.toRingedSpace.basicOpen f := by
+  rw [X.toRingedSpace.mem_basicOpen f x, ← not_iff_not, not_not]
+  exact (LocalRing.residue_ne_zero_iff_isUnit _)
+
+lemma evaluation_ne_zero_iff_mem_basicOpen (x : U) (f : X.presheaf.obj (op U)) :
+    X.evaluation x f ≠ 0 ↔ x.val ∈ X.toRingedSpace.basicOpen f := by
+  simp
+
+@[simp]
+lemma Γevaluation_eq_zero_iff_not_mem_basicOpen (x : X) (f : X.presheaf.obj (op ⊤)) :
+    X.Γevaluation x f = 0 ↔ x ∉ X.toRingedSpace.basicOpen f :=
+  evaluation_eq_zero_iff_not_mem_basicOpen X ⟨x, show x ∈ ⊤ by trivial⟩ f
+
+lemma Γevaluation_ne_zero_iff_mem_basicOpen (x : X) (f : X.presheaf.obj (op ⊤)) :
+    X.Γevaluation x f ≠ 0 ↔ x ∈ X.toRingedSpace.basicOpen f :=
+  evaluation_ne_zero_iff_mem_basicOpen X ⟨x, show x ∈ ⊤ by trivial⟩ f
+
+variable {X Y : LocallyRingedSpace.{u}} (f : X ⟶ Y)
+
+/-- If `X ⟶ Y` is a morphism of locally ringed spaces and `x` a point of `X`, we obtain
+a morphism of residue fields in the other direction. -/
+def evaluationMap (x : X) : Y.residueField (f.val.base x) ⟶ X.residueField x :=
```

Comment:
Could you also include a lemma which relates `stalkMap f x` and `evaluationMap f x`, and lemmas `evaluationMap_id` and `evaluationMap_comp`?

Reply:
(Also, I am not sure about the name "evaluationMap"; do not you feel that `mapResidueField` would be better?)

Reply:
`residueFieldMap` (in accordance with `stalkMap`) maybe?